stablehlo/reference/Index.h

```diff
@@ -0,0 +1,59 @@
+/* Copyright 2022 The StableHLO Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef STABLEHLO_REFERENCE_INDEX_H_
+#define STABLEHLO_REFERENCE_INDEX_H_
+
+#include <cstdint>
+#include <vector>
+
+#include "llvm/ADT/ArrayRef.h"
+
+namespace mlir {
+namespace stablehlo {
+
+/// Iterator over the indices of a tensor in major-to-minor order. As an
+/// example, for a tensor with shape [2,3], the iterator enumerates the
+/// indices (0,0), (0,1), (0,2), (1,0), (1,1), and (1,2).
+class IndexSpaceIterator {
+ public:
+  /// \name Constructors
+  explicit IndexSpaceIterator(const std::vector<int64_t> &shape,
+                              bool lastIndex = false);
```

Comment:
`lastIndex` is a bit awkward because it's ambiguous with respect to what it means when it's `false`. Does this mean first index? Some other index? We can avoid this ambiguity by passing the `indices` directly.

Reply:
Please see my comment above.